Mathlib/Data/Matroid/Circuit.lean

```diff
@@ -319,4 +336,128 @@ lemma ext_iff_circuit {M₁ M₂ : Matroid α} :
+`f ∈ C₁ \ C₂`, there is a circuit `C` with `f ∈ C ⊆ (C₁ ∪ C₂) \ {e}`. -/
+lemma Circuit.strong_elimination (hC₁ : M.Circuit C₁) (hC₂ : M.Circuit C₂) (heC₁ : e ∈ C₁)
+    (heC₂ : e ∈ C₂) (hfC₁ : f ∈ C₁) (hfC₂ : f ∉ C₂) :
+    ∃ C ⊆ (C₁ ∪ C₂) \ {e}, M.Circuit C ∧ f ∈ C := by
+  obtain ⟨C, hCs, hC, hfC⟩ := hC₁.strong_multi_elimination (fun i : Unit ↦ e) (fun _ ↦ C₂)
+    (by simpa) (z := f) (by simpa) (by simpa) (by simp) (by simpa) (by simpa)
+  exact ⟨C, hCs.trans (diff_subset_diff (by simp) (by simp)), hC, hfC⟩
+
+/-- The circuit elimination axiom : for any pair of distinct circuits `C₁,C₂` and any `e`, some
+circuit is contained in `(C₁ ∪ C₂) \ {e}`.
+
+This is one of the axioms when definining finitary matroid via circuits;
+as an axiom, it is usually stated with the extra assumption that `e ∈ C₁ ∩ C₂`. --/
+lemma Circuit.elimination (hC₁ : M.Circuit C₁) (hC₂ : M.Circuit C₂) (h : C₁ ≠ C₂) (e : α) :
+    ∃ C ⊆ (C₁ ∪ C₂) \ {e}, M.Circuit C := by
+  obtain ⟨f, hf₁, hf₂⟩ : (C₁ \ C₂).Nonempty := by
+    rw [nonempty_iff_ne_empty, Ne, diff_eq_empty]
+    exact fun hss ↦ h (hC₁.eq_of_subset_circuit hC₂ hss)
+  by_cases he₁ : e ∈ C₁
+  · by_cases he₂ : e ∈ C₂
+    · obtain ⟨C, hC, hC', -⟩ := hC₁.strong_elimination hC₂ he₁ he₂ hf₁ hf₂
+      exact ⟨C, hC, hC'⟩
+    exact ⟨C₂, subset_diff_singleton subset_union_right he₂, hC₂⟩
+  exact ⟨C₁, subset_diff_singleton subset_union_left he₁, hC₁⟩
+
+end Elimination
+
+/-! ### Finitary Matroids -/
+
+lemma Circuit.finite [Finitary M] (hC : M.Circuit C) : C.Finite := by
+  have hi := hC.dep.not_indep
+  rw [indep_iff_forall_finite_subset_indep] at hi; push_neg at hi
+  obtain ⟨J, hJC, hJfin, hJ⟩ := hi
+  rwa [← hC.eq_of_not_indep_subset hJ hJC]
+
+lemma finitary_iff_forall_circuit_finite : M.Finitary ↔ ∀ C, M.Circuit C → C.Finite := by
+  refine ⟨fun _ _ ↦ Circuit.finite, fun h ↦
+    ⟨fun I hI ↦ indep_iff_not_dep.2 ⟨fun hd ↦ ?_,fun x hx ↦ ?_⟩⟩⟩
+  · obtain ⟨C, hCI, hC⟩ := hd.exists_circuit_subset
+    exact hC.dep.not_indep <| hI _ hCI (h C hC)
+  simpa using (hI {x} (by simpa) (finite_singleton _)).subset_ground
+
+lemma exists_mem_finite_closure_of_mem_closure {Y : Set α} [M.Finitary] (he : e ∈ M.closure Y) :
+    ∃ I ⊆ Y, I.Finite ∧ M.Indep I ∧ e ∈ M.closure I := by
+  by_cases heY : e ∈ Y
+  · obtain ⟨J, hJ⟩ := M.exists_basis {e}
+    exact ⟨J, hJ.subset.trans (by simpa), (finite_singleton e).subset hJ.subset, hJ.indep,
+      by simpa using hJ.subset_closure⟩
+  obtain ⟨C, hCs, hC, heC⟩ := exists_circuit_of_mem_closure he heY
+  exact ⟨C \ {e}, by simpa, hC.finite.diff, hC.diff_singleton_indep heC,
+    hC.mem_closure_diff_singleton_of_mem heC⟩
+
+/-- In a finitary matroid, each finite set spanned by `X` is spanned by a finite independent
+subset of `X`. -/
+lemma exists_subset_finite_closure_of_subset_closure [M.Finitary] (hX : X.Finite)
+    (hXY : X ⊆ M.closure Y) : ∃ I ⊆ Y, I.Finite ∧ M.Indep I ∧ X ⊆ M.closure I := by
```

Comment:
If I'm reading this correctly, it's a finite set `X` spanned by `Y` being spanned by a subset of `Y`?

As an aside, as someone who doesn't know this theory well but is interested in it, I find docstrings like these incredibly valuable!

Reply:
Thanks for spotting that!